Optimize builders

iurly opened the pull request:
This PR adds a few optimizations to slightly speed up the build step process:
- only build the builders once
- do not try to build bpf probes if the builder does not explicitly support it
- add a builder for debian 12 bookworm

iurly committed:
builder: only build builder images once

we're running 'docker build' everytime we try to build a probe.
That's not needed, just do it once and record that it has been done.

iurly committed:
name bpf-capable builders explicitly and use that information

we're currently trying to build eBPF probes using builders which
clearly don't support it. Add '-bpf' suffix to builders which do,
and stop trying building eBPF unless the chosen builder's name
contains this suffix.

iurly committed:
add builder for debian 12 bookworm

deepskyblue86 commented on probe_builder/builder/builder_image.py:
```diff
@@ -7,11 +7,36 @@
 
 logger = logging.getLogger(__name__)
 
+# TODO here perhaps we might want to make a class?
```
pending TODO

iurly replied:
yes, in the future

deepskyblue86 replied:
I don't really like the idea of committing future purposes (either the comment either the commented out code)

deepskyblue86 commented on probe_builder/builder/builder_image.py:
```diff
@@ -7,11 +7,36 @@
+    if workspace.image_prefix:
+        # image_prefix essentially means: get this from a docker repo
+        pass
+        # TODO here we should have something like
```
and why aren't we doing it?

iurly replied:
my mistake, I don't think that's the intended use case

deepskyblue86 commented on probe_builder/builder/choose_builder.py:
```diff
@@ -113,17 +115,23 @@ def choose_gcc_dockerfile(builder_source, builder_distro, kernel_dir):
     prefix = 'Dockerfile.{}-gcc'.format(builder_distro)
-    dockerfile_versions = [Version(f[len(prefix):]) for f in os.listdir(builder_source) if f.startswith(prefix)]
-    dockerfile_versions.sort()
+    regex = re.compile('^' + re.escape(prefix) + '(?P<gccver>[0-9]+\.[0-9]+)(?P<bpf>(\-bpf)?)$')
```
update the comment above, maybe also mentioning why we have a `bpf` group that we don't use

iurly replied:
I changed the logic so to actually use that information

deepskyblue86 commented on probe_builder/builder/choose_builder.py:
```diff
@@ -113,17 +115,23 @@ def choose_gcc_dockerfile(builder_source, builder_distro, kernel_dir):
-    dockerfile_versions.sort()
+    regex = re.compile('^' + re.escape(prefix) + '(?P<gccver>[0-9]+\.[0-9]+)(?P<bpf>(\-bpf)?)$')
+    dockerfile_versions = [
+        (Version(regex.match(f).group('gccver')), f)
+        for f in os.listdir(builder_source)
+        if regex.match(f)
```
I'd like it better in a single line as it was before

deepskyblue86 commented on probe_builder/builder/distro/base_builder.py:
```diff
@@ -88,8 +88,11 @@ def build_kernel(self, workspace, probe, builder_distro, release, target):
                                                                       kernel_dir)
-        if not workspace.image_prefix:
-            builder_image.build(workspace, dockerfile, dockerfile_tag)
+        # let build() figure out if it actually needs to build or pull anything
+        builder_image.build(workspace, dockerfile, dockerfile_tag)
```
Maybe have a different name, if `build` only eventually builds? IMHO the comment makes sense only by looking at the diff...

deepskyblue86 commented on probe_builder/builder/distro/base_builder.py:
```diff
@@ -88,8 +88,11 @@ def build_kernel(self, workspace, probe, builder_distro, release, target):
+        builder_image.build(workspace, dockerfile, dockerfile_tag)
+
+        if not dockerfile.endswith('-bpf'):
+            ebpf_skip_reason = "Dockerfile {} does not end with -bpf".format(dockerfile)
```
I'd prefer a more meaningful message, since the -bpf suffix is just a convention we made up, it would have been on a different folder or something else.
"eBPF not supported by builder {}"?
"Builder {} does not support eBPF"?

iurly replied:
done

deepskyblue86 reviewed (commented):
I don't have enough context, I'll give another look at this later

iurly committed:
rework Dockerfile-bpf-capability feature

addressing review comments:
- add code comments make the logic a bit more explicit
- rework the selection algorithm
- make the bpf-capability an explicit feature returned
  by the builder choosing logic
